packages/block-library/src/audio/theme.scss

```diff
@@ -2,6 +2,6 @@
 	@include caption-style-theme();
 }
 
-.wp-block-audio {
+:where(.wp-block-audio) {
```

Comment:
Can you explain why these particular changes are necessary (block styles classnames)? I admit it's a bit hard to follow the reasoning of the whole changes.

Reply:
Yes! So this PR reduces the specificity of block global styles by wrapping their selectors in `:where()`. With that change, the margins set in these block-library files now have higher specificity than the global styles, so it would appear that the global styles margins have no effect. 

That made it necessary to reduce the specificity of these styles, as they shouldn't be allowed to override anything theme- or user-defined.

Reply:
Won't this impact third-party blocks using this kind of styles as well?

Reply:
I'm not sure I understand. Would third party blocks use core block-library styles at all?

Reply:
I mean some third-party blocks might be using `.my-block { margin: something }` and with the changes being done to theme.json style generation, the styles of block will be more specific and taking over (even if the users changes the margin in global styles for instance)

Reply:
Oh I see, yes. I'm having trouble creating a test case for that because I can't find any third-party blocks that work with global styles. I'm guessing support for global styles is something that has to be explicitly enabled in the block code? A quick search doesn't bring up any relevant documentation.

From previous testing with third-party blocks though, many of them don't use core block supports at all, so I don't think this will be much of a problem.

Reply:
What about the WooCommerce blocks? There are several of them that support global styles including margins.
Perhaps the Woo developers can share insights and help test.

Reply:
I can share that in custom block development for clients we use block supports very heavily. 

The CSS we write just uses the class of the block on it's own (`.wp-block-namespace-example`) to add base styling and then do expect any user preferences in either global styles or block instances to override our base styles

Reply:
Thanks for the tips folks! I tested this PR with [WooCommerce blocks ](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks) and everything seems to be working well. Browsing the block code it looks like blocks using supports don't generally set styles that could conflict (e.g. a block with color support doesn't set `color` or `background-color` but relies on inheritance to match its context).

@fabiankaegy do you have concrete instances of blocks that provide both block supports _as well as_ CSS for the supported properties? I'm wondering how common a case that might be. 

My concern is that the only viable alternative to the approach in this PR is using CSS layers to enforce a hierarchy in styles. But that idea has had pushback from folks who are concerned about legacy browser support, as @cbirdsong mentioned above. It would be better if we could go with just reducing specificity for the global styles selectors for now.

Reply:
Looking at the caniuse table for CSS layers, it seems it's already well supported 94% overall users and 100% of the WP supported browsers. So technically nothing stops us from trying it. In fact it seems like this is the prefect use-case for it. But I'm also wondering if it might have its own pitfalls (which we could discover when implementing it). I personally think it's worth looking into it, if it's not too complex.

It doesn't mean we shouldn't consider this alternative as well (or both), but I'm not sure legacy browser support is a valid concern anymore for CSS layers.

Reply:
CSS layers sound like a worthwhile exploration 👍 

My question on that approach is how would we handle making sure 3rd party enqueued stylesheets are within an appropriate layer? That would have to be handled automatically rather than left to custom blocks and plugins. 

I hope I'm missing something but would a CSS layer approach require moving from enqueuing styles via `link` and `style` elements, to something where they are all imported to a layer via `@import`? If possible, what would the performance hit there be?

Layers do seem a bit like an all or nothing approach given unlayered styles would take precedence regardless of order over layered styles.

Reply:
It is possible to do this `@import url('example.css') layer(<layer-name>);` to layer an entire stylesheet. So worth exploring whether we can update the "link" of the stylesheets that we want to auto-later with that. There's a proposal to add `layer` directly to `<link` but it's not finalized yet.

Reply:
As I mentioned in [my comment above](https://github.com/WordPress/gutenberg/pull/57841#issuecomment-1918368481), I think an initial implementation of layers could include only core CSS (such as the block library styles) and generated CSS that's meant to be low specificity (such as global styles and block supports). 

User-defined styles on specific blocks should be excluded from this as we always want them to override anything else, and theme/plugin enqueued stylesheets should almost always override core styles. I guess an exception here might be third party block stylesheets: should they be treated like core block stylesheets?

I'm not sure we should be using `@import` for stylesheets at all because it forces stylesheets to be loaded sequentially, which can have quite an impact on performance. There are some workarounds but I'm not sure any of them are 100% safe 😅

Reply:
It does seem like regardless of the outcome of the current PR, we should be exploring CSS layers. That said, it's clear that it's a big undertaking, so if you all think we should move forward with the current PR, I'm fine with that. 

Let's make sure we keep an eye on the feedback and the impact though.

Reply:
I also agree that we should definitly investigate CSS Layers. The thing I want to call out though that has always prevented me from persuing it further is that fudamentally there is no great fallback for browsers that don't support layers. 

We do have pretty decent browser supoport these days. But I think there is a big difference between features that fall back easily such as for example `text-wrap: balance`. If the brower doesn't support that, it just falls back to use another value.

If the browser doesn't support layers or container queries etc it cannot fallback to anything really and so the outcome is a very bad one.

Reply:
I do think we should move forward with the current PR regardless of any investigation into layers (which is fine to do in principle - we should try new stuff out - but I suspect the fact that at some point we'll have to rely on `@import`ing third party stylesheets will prove a blocker to actually shipping it)

The potential breaking changes will be similar whether we use layers or not, so I think it would be good to come to an agreement on style hierarchy regardless of the approach. Most changes in this PR have so far been uncontroversial; the one unresolved issue is how to manage third party block styles vs global styles.

As I said above, I'm not sure this will be a major issue because the only conflict we've found so far is with blocks that: 
* provide their own styles and simultaneously offer global styles support AND
* provide default styles in the block stylesheet **for the same properties** that can be changed in global styles

In my testing, I couldn't find any blocks that did both these things.

In any case, this will have to be clearly announced as a breaking change so block devs have the chance to make any necessary tweaks to their plugins in advance.